### PR TITLE
Create initial configuration for authz and create store as part of migration

### PR DESCRIPTION
**Pull request by JAORMX**
This ensures that we create an authz store as part of the migration command.

To enable this, an Authz configuration section was created.

The next step is to actually populate the model.

### CODE REVIEW

**Review by evankanderson — COMMENTED** [review]
What does a client call here look like?  Feel free to just free-form an example in the initial PR comment.

How do we configure the server in dev (`docker-compose`) and staging/production (checked-in configuration) to use the right parameters, particularly if we need to look up the store Id from a name?  It seems like something that we'd want `GetClient()` to do once and cache the result, for example.

**Comment by evankanderson** [comment]
With respect to populating the model, I wonder whether we want to be able to start up without an OpenFGA, create a few resources, and then run the "back-populate the model and turn on FGA authorization".  (I'm not sure, but it seems like it might enable more continuous testing of the reconciliation between the two.)

File: internal/authz/authz.go [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+}
+
+// NewAuthzClient returns a new AuthzClientWrapper
+func NewAuthzClient(cfg *srvconfig.AuthzConfig) (*ClientWrapper, error) {
+	if err := cfg.Validate(); err != nil {
+		return nil, err
+	}
+
+	credsCfg := credentials.Credentials{}
+	if cfg.Auth.Method == "token" {
```
[the remark this answers is not in the document]
Comment (JAORMX) [reply internal/authz/authz.go:43]: We already do through the Validate function

File: internal/authz/authz.go [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+		tok, err := cfg.Auth.Token.ReadToken()
```
[the remark this answers is not in the document]
Comment (JAORMX) [reply internal/authz/authz.go:44]: That's a good idea

File: cmd/server/app/migrate_up.go [hunk 3]
```diff
@@ -99,6 +101,38 @@ var upCmd = &cobra.Command{
...
+		}
+
+		var storeExists bool
+		// TODO: We might want to handle pagination here.
+		for _, store := range stores.Stores {
+			if store.Name == cfg.Authz.StoreName {
+				storeExists = true
+				break
+			}
+		}
```
[the remark this answers is not in the document]
Comment (JAORMX) [reply cmd/server/app/migrate_up.go:125]: You need to create a store and use the UUID of the previously created store

**Comment by JAORMX** [comment]
> What does a client call here look like?  Feel free to just free-form an example in the initial PR comment.
> 
> How do we configure the server in dev (`docker-compose`) and staging/production (checked-in configuration) to use the right parameters, particularly if we need to look up the store Id from a name?  It seems like something that we'd want `GetClient()` to do once and cache the result, for example.

I was hoping to iterate on several PRs and not give all the answers here, but ok. This PR is not a complete solution for bootstrapping openfga, but it's a start. We still need to add the store and model IDs.

The store ID is tricky since you need to pre create it. So it would be so two step process, first do the migration, then get the ID to take it into use. IMO we need to be able to work without a store ID, and be able to set it afterwards.

If you're interested in helping out, I'd use your input on the authentication pieces, there are client flow credentials which we could leverage, and i think that's the intended flow for the oidc auth, but it's not well documented

File: cmd/server/app/migrate_up.go [hunk 4]
```diff
@@ -99,6 +101,38 @@ var upCmd = &cobra.Command{
...
+
+		if !storeExists {
+			cmd.Printf("Creating authz store %s\n", cfg.Authz.StoreName)
+			st, err := authzcli.CreateStore(ctx).Body(fgaclient.ClientCreateStoreRequest{Name: cfg.Authz.StoreName}).Execute()
```
[the remark this answers is not in the document]
Comment (JAORMX) [reply cmd/server/app/migrate_up.go]: `ApiCreateStoreRequest` != `SdkClientCreateStoreRequestInterface`

This confused me as well, but the interface we're using does not return 3 values.

**Review by eleftherias — COMMENTED** [review]
Could we add a temporary guard around the new logic, so that it only runs if `authz` is set in the configuration?
We can remove it when we have the local development strategy figured out.
I'm worried that we're breaking `migrateup` for anyone trying to run locally.

**Comment by JAORMX** [comment]
@eleftherias folks trying to run locally would hit the OpenFGA container that's spawned via docker-compose. So that should just work. Or were you thinking of other types of deployments?

**Comment by eleftherias** [comment]
> @eleftherias folks trying to run locally would hit the OpenFGA container that's spawned via docker-compose. So that should just work. Or were you thinking of other types of deployments?

My bad, I didn't realize we already had OpenFGA in our docker-compose.

When running `make run-docker` locally, my migration fails with the error: `Error: error while creating authz client: Key: 'AuthzConfig.Auth.Token.TokenPath' Error:Field validation for 'TokenPath' failed on the 'file' tag`, because I don't have the file `/var/run/secrets/kubernetes.io/serviceaccount/token`.  

It seems it expects that file to be present even if we set `method: none`.

**Comment by JAORMX** [comment]
@eleftherias lol I just added that validation in that last PR. Seems I have a bug in the validation then. thanks for spotting it! I thought it would be caught by our migration test which also spawns docker-compose.

**Comment by JAORMX** [comment]
@eleftherias Ah! the issue was that the default is `token`. I switched it to `none` since that's what would work with the local deployment.

**Comment by evankanderson** [comment]
> @eleftherias Ah! the issue was that the default is `token`. I switched it to `none` since that's what would work with the local deployment.

Yes, I'd start with `auth: none` (and the enable/disable switch) to get other stuff like the store ID working, then circle back to the auth params (which I'm willing to investigate).

**Comment by evankanderson** [comment]
Again, I like the idea of starting small.  IMO, the following would be sufficient:

1. Some sort of flow for getting StoreID -- if we're creating the store in migrate_up (if it's not provided?), then there needs to be some way for the server process to find the store ID.
2. Some sort of mechanism (library call or context hook, depending on what's needed) which exposes an interface that components can call to get an authorization decision.  As a wild SWAG:
   ```golang
   type Authz interface {
      Check(ctx context.Context) fgaclient.SdkClientCheckRequestInterface
      CheckExecute(request fgaclient.SdkClientCheckRequestInterface) (*fgaclient.ClientCheckResponse, error)
   }
   ```
   And we could add `Write`, `Delete`, etc later.  Alternatively, if you don't like the `SdkClient` interface, you could just have `Check(ctx context.Context, request fgaclient.SdkClientCheckRequestInterface) (*fgaclient.ClientCheckResponse, error)`, and I wouldn't bat an eye at that, either.

**Comment by JAORMX** [comment]
> Some sort of flow for getting StoreID -- if we're creating the store in migrate_up (if it's not provided?), then there needs to be some way for the server process to find the store ID.


Was thinking of just passing it in in a subsequent run. But having the server work by listing the stores anyway.

**Comment by JAORMX** [comment]
> Some sort of mechanism (library call or context hook, depending on what's needed) which exposes an interface that components can call to get an authorization decision. As a wild SWAG:

I agree, though I was thinking of adding that in a subsequent PR

**Comment by JAORMX** [comment]
I'm not sure how to go forward with this. I wanted to keep small iterations going and leave the authorization logic, model setting, and further details in subsequent PRs and I'm not sure we're in the same page in regards to the review. Would you prefer I add more aspects to this PR?

**Review by evankanderson — DISMISSED** [review]
We're not going to focus on parallelization, so let's get this in.

**Review by eleftherias — APPROVED** [review]

**Pull request merged**